commit 51f554f909086f4baca7be215d5edacab744dea4
Author: Kartikaya Gupta <kats@calvin.staktrace.com>
Date:   Wed Mar 7 23:30:48 2012 -0500

    Modifications needed to get build working on Mac OS X 10.6

diff --git a/.gitignore b/.gitignore
new file mode 100644
index 0000000..3e90033
--- /dev/null
+++ b/.gitignore
@@ -0,0 +1,6 @@
+*.o
+Makefile
+libbitdht/src/lib/
+libretroshare/src/lib/
+retroshare-gui/src/RetroShare.app/
+retroshare-gui/src/temp/
diff --git a/libretroshare/src/libretroshare.pro b/libretroshare/src/libretroshare.pro
index 6f35e01..267cc91 100644
--- a/libretroshare/src/libretroshare.pro
+++ b/libretroshare/src/libretroshare.pro
@@ -312,7 +312,7 @@ mac {
 		OBJECTS_DIR = temp/obj
 		MOC_DIR = temp/moc
 		#DEFINES = WINDOWS_SYS WIN32 STATICLIB MINGW
-                #DEFINES *= MINIUPNPC_VERSION=13
+		DEFINES *= MINIUPNPC_VERSION=13
 		DESTDIR = lib
 
                 #miniupnp implementation files
@@ -326,12 +326,11 @@ mac {
 		# Beautiful Hack to fix 64bit file access.
                 QMAKE_CXXFLAGS *= -Dfseeko64=fseeko -Dftello64=ftello -Dfopen64=fopen -Dvstatfs64=vstatfs
 
-                UPNPC_DIR = ../../../miniupnpc-1.0
-		GPG_ERROR_DIR = ../../../../libgpg-error-1.7
-		GPGME_DIR  = ../../../../gpgme-1.1.8
+		UPNPC_DIR = ../../../miniupnpc-1.3
+		GPG_ERROR_DIR = ../../../libgpg-error-1.10
+		GPGME_DIR  = ../../../gpgme-1.3.1
 
-		INCLUDEPATH += . $${UPNPC_DIR} 
-		#INCLUDEPATH += . $${UPNPC_DIR} $${GPGME_DIR}/src $${GPG_ERROR_DIR}/src
+		INCLUDEPATH += . $${UPNPC_DIR} $${GPGME_DIR}/src $${GPG_ERROR_DIR}/src
 }
 
 ################################# FreeBSD ##########################################
diff --git a/libretroshare/src/pqi/sslfns.cc b/libretroshare/src/pqi/sslfns.cc
index 2588cb8..305433f 100644
--- a/libretroshare/src/pqi/sslfns.cc
+++ b/libretroshare/src/pqi/sslfns.cc
@@ -586,8 +586,8 @@ X509 *loadX509FromDER(const uint8_t *ptr, uint32_t len)
 	X509 *tmp = NULL;
 #ifdef __APPLE__
 	// This depends on which version you are compiling for... OSX10.5 doesn't have consts (old OpenSSL!)
-	unsigned char **certptr = (unsigned char **) &ptr;
-	//const unsigned char **certptr = (const unsigned char **) &ptr;
+	//unsigned char **certptr = (unsigned char **) &ptr;
+	const unsigned char **certptr = (const unsigned char **) &ptr;
 #else
 	const unsigned char **certptr = (const unsigned char **) &ptr;
 #endif
diff --git a/retroshare-gui/src/RetroShare.pro b/retroshare-gui/src/RetroShare.pro
index adcb5a0..ffbf669 100644
--- a/retroshare-gui/src/RetroShare.pro
+++ b/retroshare-gui/src/RetroShare.pro
@@ -140,19 +140,21 @@ win32 {
 
 macx {
     # ENABLE THIS OPTION FOR Univeral Binary BUILD.
-    	CONFIG += ppc x86
-	QMAKE_MACOSX_DEPLOYMENT_TARGET = 10.4
+	# CONFIG += ppc x86
+	QMAKE_MACOSX_DEPLOYMENT_TARGET = 10.6
 
 	CONFIG += version_detail_bash_script
 	LIBS += ../../libretroshare/src/lib/libretroshare.a
-        LIBS += -lssl -lcrypto -lz -lgpgme -lgpg-error -lassuan
-	LIBS += ../../../miniupnpc-1.0/libminiupnpc.a
+	LIBS += ../../../libgpg-error-1.10/build/lib/libgpg-error.a
+	LIBS += ../../../gpgme-1.3.1/build/lib/libgpgme.a
+	LIBS += ../../../miniupnpc-1.3/libminiupnpc.a
+	LIBS += -lssl -lcrypto -lz -lassuan
 	LIBS += -framework CoreFoundation
 	LIBS += -framework Security
 
  #	LIBS += -framework CoreServices
 
-    	INCLUDEPATH += .
+	INCLUDEPATH += . ../../../gpgme-1.3.1/src
 	#DEFINES* = MAC_IDLE # for idle feature
 	CONFIG -= uitools
 
